File: python/0095.py
# Link: https://leetcode.com/problems/unique-binary-search-trees-ii/description/
# Runtime: 52 ms
# Memory: 17.5 MB

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.__generate_all_trees(1, n)
    
    @cache
    def __generate_all_trees(self, left: int, right: int) -> List[Optional[TreeNode]]:
        if left > right: return [None]
        elif left == right: return [TreeNode(left)]

        all_trees = []
        for val in range(left, right + 1):
            for left_tree in self.__generate_all_trees(left, val - 1):
                for right_tree in self.__generate_all_trees(val + 1, right):
                    all_trees.append(TreeNode(val, left_tree, right_tree))
        
        return all_trees

File: c++/0095.cpp
// Link: https://leetcode.com/problems/unique-binary-search-trees-ii/description/
// Runtime: 9 ms
// Memory: 12.52 MB

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    std::vector<TreeNode*> generateTrees(int n) 
    {
        return generateAllTrees(1, n);
    }

private:
    std::map<std::pair<int, int>, std::vector<TreeNode*>> subtrees;

    std::vector<TreeNode*> generateAllTrees(int left, int right)
    {
        if (left > right) return {nullptr};
        else if (left == right) return {new TreeNode(left)};
        else if (subtrees.find({left, right}) != subtrees.end()) return subtrees[{left, right}];

        std::vector<TreeNode*> allTrees;
        for (int i = left; i < right + 1; i++)
            for (const auto& leftTree : generateAllTrees(left, i - 1))
                for (const auto& rightTree : generateAllTrees(i + 1, right))
                    allTrees.push_back(new TreeNode(i, leftTree, rightTree));

        subtrees[{left, right}] = allTrees;            
        
        return allTrees;
    }
};
